azure-eventhubs-eph/src/main/java/com/microsoft/azure/eventprocessorhost/PartitionManager.java

```diff
@@ -95,339 +123,370 @@ void onPartitionCheckCompleteTestHook()
+        // Stage 0: does lease store exist?
+        CompletableFuture<?> initializeStoresFuture = leaseManager.leaseStoreExists()
+        // Stage 1: create lease store if it doesn't exist
+        .thenComposeAsync((exists) ->
         {
-        	retryWrapper(() -> checkpointManager.createCheckpointStoreIfNotExists(), null, "Failure creating checkpoint store for this Event Hub, retrying",
-        			"Out of retries creating checkpoint store for this Event Hub", EventProcessorHostActionStrings.CREATING_CHECKPOINT_STORE, 5);
-        }
-        // else
-        //	checkpoint store already exists, no work needed
+        	CompletableFuture<?> createLeaseStoreFuture = null;
+        	if (!exists)
+        	{
+        		createLeaseStoreFuture = retryWrapper(() -> leaseManager.createLeaseStoreIfNotExists(), null,
+        				"Failure creating lease store for this Event Hub, retrying",
+						"Out of retries creating lease store for this Event Hub", EventProcessorHostActionStrings.CREATING_LEASE_STORE, 5);
+        	}
+        	else
+        	{
+        		createLeaseStoreFuture = CompletableFuture.completedFuture(null);
+        	}
+        	return createLeaseStoreFuture;
+        }, this.host.getExecutorService())
+        // Stage 2: does checkpoint store exist?
+        .thenComposeAsync((unused) -> checkpointManager.checkpointStoreExists(), this.host.getExecutorService())
+        // Stage 3: create checkpoint store if it doesn't exist
+        .thenComposeAsync((exists) ->
+        {
+        	CompletableFuture<?> createCheckpointStoreFuture = null;
+        	if (!exists)
+        	{
+        		createCheckpointStoreFuture = retryWrapper(() -> checkpointManager.createCheckpointStoreIfNotExists(), null,
+        				"Failure creating checkpoint store for this Event Hub, retrying",
+            			"Out of retries creating checkpoint store for this Event Hub", EventProcessorHostActionStrings.CREATING_CHECKPOINT_STORE, 5);
+        	}
+        	else
+        	{
+        		createCheckpointStoreFuture = CompletableFuture.completedFuture(null);
+        	}
+        	return createCheckpointStoreFuture;
+        }, this.host.getExecutorService());
         
-        // Now make sure the checkpoints exist
-        for (String id : getPartitionIds())
+        // Stages 4-n: by now, either the stores exist (stages 0-3 succeeded) or one of them completed exceptionally and
+        // all these stages will be skipped
+        for (String id : this.partitionIds)
         {
-        	retryWrapper(() -> checkpointManager.createCheckpointIfNotExists(id), id, "Failure creating checkpoint for partition, retrying",
-        			"Out of retries creating checkpoint blob for partition", EventProcessorHostActionStrings.CREATING_CHECKPOINT, 5);
+        	final String iterationId = id;
+        	initializeStoresFuture = initializeStoresFuture
+        	// Stage X: create lease for partition <iterationId>
+        	.thenComposeAsync((unused) ->
```

Comment:
>thenComposeAsync [](start = 10, length = 16)

Will tasks be scheduled one at a time in this pattern? Or, all of them are scheduled and executed sequentially?

Reply:
This is a good point. The intent was for retryWrapper to build a set of stages that would be executed sequentially and composed into the chain that initializeStores is building. Based on watching the CITs run, that seems to be what actually happens, but the code doesn't express the intent very well. A simple code change will make it more obvious.